File: app/src/main/java/com/kickstarter/ui/activities/LoginActivity.java
package com.kickstarter.ui.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Html;
import android.util.Pair;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.kickstarter.KSApplication;
import com.kickstarter.R;
import com.kickstarter.libs.ActivityRequestCodes;
import com.kickstarter.libs.BaseActivity;
import com.kickstarter.libs.KSString;
import com.kickstarter.libs.qualifiers.RequiresActivityViewModel;
import com.kickstarter.libs.utils.ObjectUtils;
import com.kickstarter.libs.utils.ViewUtils;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.toolbars.LoginToolbar;
import com.kickstarter.ui.views.ConfirmDialog;
import com.kickstarter.viewmodels.LoginViewModel;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.BindString;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import rx.Observable;

import static com.kickstarter.libs.rx.transformers.Transformers.observeForUI;
import static com.kickstarter.libs.utils.TransitionUtils.slideInFromLeft;

@RequiresActivityViewModel(LoginViewModel.class)
public final class LoginActivity extends BaseActivity<LoginViewModel> {
  private ConfirmDialog confirmResetPasswordSuccessDialog;

  protected @Bind(R.id.email) EditText emailEditText;
  protected @Bind(R.id.forgot_your_password_text_view) TextView forgotPasswordTextView;
  protected @Bind(R.id.login_button) Button loginButton;
  protected @Bind(R.id.login_toolbar) LoginToolbar loginToolbar;
  protected @Bind(R.id.password) EditText passwordEditText;

  protected @BindString(R.string.login_buttons_forgot_password_html) String forgotPasswordString;
  protected @BindString(R.string.forgot_password_we_sent_an_email_to_email_address_with_instructions_to_reset_your_password) String forgotPasswordSentEmailString;
  protected @BindString(R.string.login_errors_does_not_match) String loginDoesNotMatchString;
  protected @BindString(R.string.login_errors_unable_to_log_in) String unableToLoginString;
  protected @BindString(R.string.login_buttons_log_in) String loginString;
  protected @BindString(R.string.login_errors_title) String errorTitleString;

  @Inject KSString ksString;

  @Override
  protected void onCreate(final @Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    setContentView(R.layout.login_layout);
    ((KSApplication) getApplication()).component().inject(this);
    ButterKnife.bind(this);
    this.loginToolbar.setTitle(this.loginString);
    this.forgotPasswordTextView.setText(Html.fromHtml(this.forgotPasswordString));

    errorMessages()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(e -> ViewUtils.showDialog(this, this.errorTitleString, e));

    this.viewModel.errors.tfaChallenge()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(__ -> startTwoFactorActivity());

    this.viewModel.outputs.loginSuccess()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(__ -> onSuccess());

    this.viewModel.outputs.prefillEmailFromPasswordReset()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(this.emailEditText::setText);

    this.viewModel.outputs.showResetPasswordSuccessDialog()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(showAndEmail -> {
        final boolean show = showAndEmail.first;
        final String email = showAndEmail.second;
        if (show) {
          resetPasswordSuccessDialog(email).show();
        } else {
          resetPasswordSuccessDialog(email).dismiss();
        }
      });

    this.viewModel.outputs.setLoginButtonIsEnabled()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(this::setLoginButtonEnabled);
  }

  /**
   * Lazily creates a reset password success confirmation dialog and stores it in an instance variable.
   */
  private @NonNull ConfirmDialog resetPasswordSuccessDialog(final @NonNull String email) {
    if (this.confirmResetPasswordSuccessDialog == null) {
      final String message = this.ksString.format(this.forgotPasswordSentEmailString, "email", email);
      this.confirmResetPasswordSuccessDialog = new ConfirmDialog(this, null, message);

      this.confirmResetPasswordSuccessDialog
        .setOnDismissListener(__ -> this.viewModel.inputs.resetPasswordConfirmationDialogDismissed());
      this.confirmResetPasswordSuccessDialog
        .setOnCancelListener(__ -> this.viewModel.inputs.resetPasswordConfirmationDialogDismissed());
    }
    return this.confirmResetPasswordSuccessDialog;
  }

  private Observable<String> errorMessages() {
    return this.viewModel.errors.invalidLoginError()
      .map(ObjectUtils.coalesceWith(this.loginDoesNotMatchString))
      .mergeWith(this.viewModel.errors.genericLoginError()
        .map(ObjectUtils.coalesceWith(this.unableToLoginString))
      );
  }

  @Override
  protected void onActivityResult(final int requestCode, final int resultCode, final @Nullable Intent intent) {
    super.onActivityResult(requestCode, resultCode, intent);

    if (requestCode != ActivityRequestCodes.LOGIN_FLOW) {
      return;
    }

    setResult(resultCode, intent);
    finish();
  }

  @OnTextChanged(R.id.email)
  void onEmailTextChanged(final @NonNull CharSequence email) {
    this.viewModel.inputs.email(email.toString());
  }

  @OnTextChanged(R.id.password)
  void onPasswordTextChanged(final @NonNull CharSequence password) {
    this.viewModel.inputs.password(password.toString());
  }

  @OnClick(R.id.forgot_your_password_text_view)
  public void forgotYourPasswordTextViewClick() {
    final Intent intent = new Intent(this, ResetPasswordActivity.class);
    startActivityWithTransition(intent, R.anim.slide_in_right, R.anim.fade_out_slide_out_left);
  }

  @OnClick(R.id.login_button)
  public void loginButtonOnClick() {
    this.viewModel.inputs.loginClick();
  }

  public void onSuccess() {
    setResult(Activity.RESULT_OK);
    finish();
  }

  public void setLoginButtonEnabled(final boolean enabled) {
    this.loginButton.setEnabled(enabled);
  }

  public void startTwoFactorActivity() {
    final Intent intent = new Intent(this, TwoFactorActivity.class)
      .putExtra(IntentKey.EMAIL, this.emailEditText.getText().toString())
      .putExtra(IntentKey.PASSWORD, this.passwordEditText.getText().toString());
    startActivityForResult(intent, ActivityRequestCodes.LOGIN_FLOW);
    overridePendingTransition(R.anim.slide_in_right, R.anim.fade_out_slide_out_left);
  }

  @Override
  protected @Nullable Pair<Integer, Integer> exitTransition() {
    return slideInFromLeft();
  }
}

File: app/src/main/java/com/kickstarter/ui/activities/LoginActivity.kt
package com.kickstarter.ui.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Pair
import androidx.activity.ComponentActivity
import androidx.activity.OnBackPressedCallback
import androidx.activity.compose.setContent
import androidx.activity.viewModels
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material.rememberScaffoldState
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rxjava2.subscribeAsState
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import com.kickstarter.R
import com.kickstarter.libs.ActivityRequestCodes
import com.kickstarter.libs.KSString
import com.kickstarter.libs.featureflag.FlagKey
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.coalesceWithV2
import com.kickstarter.libs.utils.extensions.getEnvironment
import com.kickstarter.libs.utils.extensions.getResetPasswordIntent
import com.kickstarter.ui.IntentKey
import com.kickstarter.ui.activities.compose.login.LoginScreen
import com.kickstarter.ui.compose.designsystem.KickstarterApp
import com.kickstarter.ui.data.ActivityResult.Companion.create
import com.kickstarter.ui.data.LoginReason
import com.kickstarter.ui.extensions.finishWithAnimation
import com.kickstarter.ui.extensions.startDisclaimerActivity
import com.kickstarter.viewmodels.LoginViewModel
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable

class LoginActivity : ComponentActivity() {

    private lateinit var ksString: KSString

    private val resetPasswordSentEmailString =
        R.string.forgot_password_we_sent_an_email_to_email_address_with_instructions_to_set_your_password
    private val loginDoesNotMatchString = R.string.login_errors_does_not_match
    private val unableToLoginString = R.string.login_errors_unable_to_log_in

    private lateinit var viewModelFactory: LoginViewModel.Factory
    private val viewModel: LoginViewModel.LoginViewModel by viewModels { viewModelFactory }

    private lateinit var disposables: CompositeDisposable

    private var darkModeEnabled = false
    private var currentEmail = ""
    private var currentPassword = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        disposables = CompositeDisposable()

        val env = this.getEnvironment()?.let { env ->
            viewModelFactory = LoginViewModel.Factory(env, intent = intent)
            darkModeEnabled =
                env.featureFlagClient()?.getBoolean(FlagKey.ANDROID_DARK_MODE_ENABLED) ?: false
            env
        }

        setContent {
            var prefillEmail = viewModel.outputs.prefillEmail().subscribeAsState(initial = "").value

            var isLoading = viewModel.isLoading().subscribeAsState(initial = false).value

            var error by remember { mutableStateOf("") }

            errorMessages().subscribe {
                error = it
            }.addToDisposable(disposables)

            var scaffoldState = rememberScaffoldState()

            var resetPasswordDialogMessage by rememberSaveable { mutableStateOf("") }

            var showDialog by rememberSaveable { mutableStateOf(false) }

            var showResetPasswordDialog = viewModel.outputs.showResetPasswordSuccessDialog()
                .subscribeAsState(initial = Pair(false, Pair("", LoginReason.DEFAULT))).value
            when {
                showResetPasswordDialog.first -> {
                    val emailAndReason = showResetPasswordDialog.second
                    val message =
                        if (emailAndReason.second == LoginReason.RESET_FACEBOOK_PASSWORD) {
                            this.ksString.format(
                                getString(this.resetPasswordSentEmailString),
                                "email",
                                emailAndReason.first
                            )
                        } else {
                            this.ksString.format(
                                getString(this.resetPasswordSentEmailString),
                                "email",
                                emailAndReason.first
                            )
                        }
                    resetPasswordDialogMessage = message
                    showDialog = true
                }

                else -> {
                    resetPasswordDialogMessage = ""
                    showDialog = false
                }
            }

            var changedPasswordMessage by remember { mutableStateOf("") }

            this.viewModel.outputs.showChangedPasswordSnackbar()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe {
                    changedPasswordMessage = getString(R.string.Got_it_your_changes_have_been_saved)
                }
                .addToDisposable(disposables)

            var createPasswordMessage by remember { mutableStateOf("") }

            this.viewModel.outputs.showCreatedPasswordSnackbar()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe {
                    createPasswordMessage = getString(R.string.Got_it_your_changes_have_been_saved)
                }
                .addToDisposable(disposables)

            KickstarterApp(useDarkTheme = if (darkModeEnabled) isSystemInDarkTheme() else false) {
                LoginScreen(
                    prefillEmail = prefillEmail,
                    scaffoldState = scaffoldState,
                    isLoading = isLoading,
                    onBackClicked = { onBackPressedDispatcher.onBackPressed() },
                    onLoginClicked = { email, password ->
                        currentEmail = email
                        currentPassword = password
                        viewModel.inputs.email(email)
                        viewModel.inputs.password(password)
                        viewModel.inputs.loginClick()
                    },
                    onTermsOfUseClicked = { startDisclaimerActivity(DisclaimerItems.TERMS) },
                    onPrivacyPolicyClicked = { startDisclaimerActivity(DisclaimerItems.PRIVACY) },
                    onCookiePolicyClicked = { startDisclaimerActivity(DisclaimerItems.COOKIES) },
                    onHelpClicked = { startDisclaimerActivity(DisclaimerItems.HELP) },
                    onForgotPasswordClicked = { startResetPasswordActivity() },
                    resetPasswordDialogMessage = resetPasswordDialogMessage,
                    showDialog = showDialog,
                    setShowDialog = {
                        showDialog = it
                        this.viewModel.inputs.resetPasswordConfirmationDialogDismissed()
                    }
                )
            }

            when {
                error.isNotEmpty() -> {
                    LaunchedEffect(scaffoldState) {
                        scaffoldState.snackbarHostState.showSnackbar(error, actionLabel = "error")
                        error = ""
                    }
                }

                changedPasswordMessage.isNotEmpty() -> {
                    LaunchedEffect(scaffoldState) {
                        scaffoldState.snackbarHostState.showSnackbar(
                            changedPasswordMessage,
                            actionLabel = "success"
                        )
                        changedPasswordMessage = ""
                    }
                }

                createPasswordMessage.isNotEmpty() -> {
                    LaunchedEffect(scaffoldState) {
                        scaffoldState.snackbarHostState.showSnackbar(
                            createPasswordMessage,
                            actionLabel = "success"
                        )
                        createPasswordMessage = ""
                    }
                }
            }
        }

        this.ksString = requireNotNull(env?.ksString())

        this.viewModel.outputs.tfaChallenge()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { startTwoFactorActivity() }
            .addToDisposable(disposables)

        this.viewModel.outputs.loginSuccess()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { onSuccess() }
            .addToDisposable(disposables)

        onBackPressedDispatcher.addCallback(object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                this@LoginActivity.finishWithAnimation()
            }
        })
    }

    private fun errorMessages() =
        this.viewModel.outputs.invalidLoginError()
            .map(coalesceWithV2(getString(this.loginDoesNotMatchString)))
            .mergeWith(
                this.viewModel.outputs.genericLoginError()
                    .map(coalesceWithV2(getString(this.unableToLoginString)))
            )

    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        super.onActivityResult(requestCode, resultCode, intent)
        viewModel.activityResult(create(requestCode, resultCode, intent))

        if (requestCode != ActivityRequestCodes.LOGIN_FLOW && requestCode != ActivityRequestCodes.RESET_FLOW) {
            return
        }

        if (requestCode != ActivityRequestCodes.RESET_FLOW) {
            setResult(resultCode, intent)
            finish()
        }
    }

    override fun onDestroy() {
        disposables.clear()
        super.onDestroy()
    }

    private fun onSuccess() {
        setResult(Activity.RESULT_OK)
        finish()
    }

    private fun startTwoFactorActivity() {
        val intent = Intent(this, TwoFactorActivity::class.java)
            .putExtra(IntentKey.EMAIL, currentEmail)
            .putExtra(IntentKey.PASSWORD, currentPassword)
        startActivityForResult(intent, ActivityRequestCodes.LOGIN_FLOW)
        overridePendingTransition(R.anim.slide_in_right, R.anim.fade_out_slide_out_left)
    }

    private fun startResetPasswordActivity() {
        val intent = Intent().getResetPasswordIntent(this, email = currentEmail)
        startActivityForResult(intent, ActivityRequestCodes.RESET_FLOW)
        overridePendingTransition(R.anim.slide_in_right, R.anim.fade_out_slide_out_left)
    }
}
